### ISSUE TITLE
cli_wallet:  0 Exception: unspecified - "Timer Expired"

### ISSUE THREAD

**Issue by D3m0nKingx** [DECENTfoundation/DECENT-Network#194]
After initiating the decentd (daemon) running on linux ubuntu  16.04 (64-bit), I started the cli_wallet, which produced the following output:

`Logging RPC to file: /home/ader/.decent/logs/rpc/rpc.log
Logging transfers to file: /home/ader/.decent/logs/transfer.log`

`Starting a new wallet with chain ID 0000000000000000000000000000000000000000000000000000000000000000 (empty one)`
`0 exception: unspecified`
`Timer Expired`
   ` {"message":"Timer Expired"}`
    `asio  websocket.cpp:656 operator()`

`    {"uri":"ws://localhost:8090"}`
   ` th_a  websocket.cpp:906 connect`

Furthermore, the rpc folder and log file do not exist, and the transfer.log file is empty

**Comment by ejossev** [comment]
Does it happen all the time even after decentd restart?

**Comment by D3m0nKingx** [comment]
yes everytime

**Comment by ejossev** [comment]
Can you telnet to localhost port 8090? To see it is daemon problem or wallet problem...

**Comment by D3m0nKingx** [comment]
ader@HP-15-Notebook-PC:~/coins/DECENT/build/artifacts/prefix/bin$ netstat -an |grep 8090

```
tcp        0      0 127.0.0.1:8090          0.0.0.0:*               LISTEN     
```

ader@HP-15-Notebook-PC:~/coins/DECENT/build/artifacts/prefix/bin$ telnet localhost 8090

```
Trying ::1...
Trying 127.0.0.1...
telnet: Unable to connect to remote host: Connection timed out
```

**Comment by D3m0nKingx** [comment]
ader@HP-15-Notebook-PC:~/coins/DECENT/build/artifacts/prefix/bin$ nmap -p 8090 localhost
```
Starting Nmap 7.01 ( https://nmap.org ) at 2017-07-02 03:24 EDT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.00034s latency).
Other addresses for localhost (not scanned): ::1
PORT     STATE    SERVICE
8090/tcp filtered unknown

Nmap done: 1 IP address (1 host up) scanned in 0.33 seconds
```

seems to be binding correctly to the port, not sure why it's timing out when connecting with telnet, as well as with the cli_wallet

**Comment by D3m0nKingx** [comment]
I think this is the problem: 
```
Proto Recv-Q Send-Q    Local Address           Foreign Address         State       PID/Program name
tcp        0      0   127.0.0.1:8090                 0.0.0.0:*        LISTEN          22482/decentd   
```
Local and Foreign Address should be the **same** on 127.0.0.1, or both separately, as it is with:
```
tcp        0      0 0.0.0.0:5297            0.0.0.0:*               LISTEN      22482/decentd   
```

I was able to successfully telnet to port 5297 of the daemon, just not 8090

**Comment by ejossev** [comment]
5297 is the p2p port, not json-rpc one. Very strange, and I can't reproduce. Can you compile the binary (Linux?) with debug info and send it to me please?

**Comment by D3m0nKingx** [comment]
Compiled it with this command:
`cmake -G "Unix Makefiles" -D CMAKE_BUILD_TYPE=Debug ~/coins/DECENT`

download decentd: https://mega.nz/#!19on3DYC!ygHICQZNV3vdwcJ-oSKXwU3KXlBNJDRPoCUXWLomTvw


What extra files/etc(please specify) do you need besides the binary?

**Comment by tigrangh** [comment]
are you able to test your 127.0.0.1 interface with two netcat instances?
have one netcat listening on 127.0.0.1:8888 and another netcat trying to connect that port.
I used to have a similar problem which is because of routing.

**Comment by D3m0nKingx** [comment]
![image](https://user-images.githubusercontent.com/18432686/27768748-e6a48f0a-5ee8-11e7-8da8-e58c4b50c05d.png)
works :100:

**Comment by D3m0nKingx** [comment]
@ejossev any luck finding the cause of the timeout ? I would like to be able to run the wallet and try the mining and seeding functionality soon.

**Comment by ejossev** [comment]
Sorry, after a while I'm spending the day with my family... It's on my tomorrow's todo list.

**Comment by D3m0nKingx** [comment]
Just for shits'n giggles I changed the rpc-endpoint in the config.ini to 0.0.0.0, however this was to of no avail :-1:

**Comment by ejossev** [comment]
Hmm, works on my virtual box with your binary. Sorry, there is no way how I can reproduce...

**Comment by D3m0nKingx** [comment]
could send me copy of your binaries (both decentd and cli_wallet)

**Comment by ejossev** [comment]
The decentd is from you... Instead of cli_wallet you can use e.g. wscat or even telnet to port 8090...

**Comment by ejossev** [comment]
But you can check my binaries as well: https://uloz.to/!5SQSpGgEQDuO/decent-tbz

**Comment by D3m0nKingx** [comment]
I just tested your binaries and they're working :+1: I have no idea why, I actually did a fresh install of Ubuntu today, and it's all good ! Maybe mine aren't compiling right ?

**Comment by D3m0nKingx** [comment]
In regards to using the cli_wallet, a more detailed instructional guide would be nice, as I'm not sure what is the necessary information needed for example the parameter <account name>, or how to use commands such as get_miner, or any other for that matter. I'm not having much success atm with doing the import_key method, just getting parsing errors, so I'm not sure what is the right information I'm supposed to provide.

**Comment by tigrangh** [comment]
"gethelp" is useful.
for example, use
"gethelp import_key"

**Comment by D3m0nKingx** [comment]
```
10 assert_exception: Assert Exception
rec && rec->name == account_name_or_id: 
    {}
    th_a  wallet.cpp:655 get_account
```
This is the result of doing import_key, is this correct output, doesn't seem so? as I'm not sure what the <accountname> for the method should be? Thank for you helping.

**Comment by D3m0nKingx** [comment]
Success at last, figured out the `<accountname>` is my encrypted wallet address. Also I compiled decent on my system, and it's working as well. Total mystery whatever was causing the localhost connection problem. thank you for all the help @ejossev @tigrangh

**Comment by D3m0nKingx** [comment]
The Timeout Error has returned again, however I believe I may have found what might be causing this, which is Ubuntu's Firewall, I'm just not sure entirely why it is.

**Comment by D3m0nKingx** [comment]
I had to actually reboot the OS after resetting / disabling the firewall/iptables, and the cli_wallet was able to connect again.

Gonna have to investigate into this, as to why and/how it's being denied with default firewall settings.

**Comment by D3m0nKingx** [comment]
@ejossev it seems someone has come across a temporary solution for connecting to the daemon on the localhost port 8090, by using a loopback socket connection through decentgo.com

`cli_wallet -s wss://socket.decentgo.com:8090`

However I'm not so sure how secure it is to be connecting with the daemon in this manner as that may not be a secure pathway. Could you please elaborate on this.

**Comment by ejossev** [comment]
Over this channel you only read from database (which has exactly the same data as anyone's local database running decentd) and send signed operations (which will appear in the blockchain sooner or later). Only sensitive part is "what queries do you run over the DB", not the content of the queries itself. E.g. anyone can see that the account kingX has 1.2345 DCTs, but the fact that you're pulling it every now and then indicates your interest in that account.

**Comment by D3m0nKingx** [comment]
The resolution to this issue has been found thanks to slack user @roelandp:

https://github.com/zaphoyd/websocketpp/commit/0bb33e4bca4ccc42a36aa2321e4fb97f2562e519

I will make a PR for this, and you can decide if to merge or wait for them to release 0.8 WebsocketPP

**Comment by roelandp** [comment]
@D3m0nKingx I am not only slack user @roelandp but also here on Github and also ~miner~ witness for Decent named @roelandp :) 

FYI the -s above is not working, you need somehow to omit the space to connect to define a custom websocket server: 

`cli_wallet -**swss**://socket.decentgo.com:8090`

It might also be of interest to read the discussion: 
https://github.com/zaphoyd/websocketpp/issues/623

**Comment by D3m0nKingx** [comment]
Fix is working, compiled and tested on Ubuntu 16.04.2 LTS x64, I have issued a PR for the fix: 
https://github.com/DECENTfoundation/websocketpp/pull/1

Again thank you @roelandp ! Happy to finally have my firewall up and running :100:

**Comment by D3m0nKingx** [comment]
Issue fixed in [4ebb31a](https://github.com/DECENTfoundation/websocketpp/commit/4ebb31a91d41b9f7b19f83a92284fc6e97d4f1bb)
